File: ports/java/src/main/java/com/gengine/Parser.java
package com.gengine;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Lexer + recursive-descent parser for the Java port. Grammar matches the
 * Python and C++ ports verbatim so the three MVPs can share rule files.
 */
public final class Parser {

    // ---------------------------------------------------------------------
    // Tokens
    // ---------------------------------------------------------------------

    enum TokKind {
        NUMBER, STRING, IDENT, OP, KEYWORD, EOF
    }

    record Token(TokKind kind, String text, Object value, int line, int col) { }

    private static final Set<String> KEYWORDS = Set.of(
            "rule", "desc", "salience", "begin", "end",
            "if", "else", "true", "false", "return"
    );

    // Precedence table for the expression parser. Missing operators
    // (assignment, comma) are handled as statement-level syntax.
    private static final Map<String, Integer> PRECEDENCE = Map.ofEntries(
            Map.entry("||", 1), Map.entry("&&", 2),
            Map.entry("==", 3), Map.entry("!=", 3),
            Map.entry("<", 3), Map.entry("<=", 3),
            Map.entry(">", 3), Map.entry(">=", 3),
            Map.entry("+", 4), Map.entry("-", 4),
            Map.entry("*", 5), Map.entry("/", 5), Map.entry("%", 5)
    );

    // ---------------------------------------------------------------------
    // Lexer
    // ---------------------------------------------------------------------

    static List<Token> tokenize(String src) {
        List<Token> out = new ArrayList<>();
        int i = 0, line = 1, lineStart = 0;
        while (i < src.length()) {
            char c = src.charAt(i);
            if (c == '\n') { line++; lineStart = ++i; continue; }
            if (Character.isWhitespace(c)) { i++; continue; }
            // Line comment "// ..."
            if (c == '/' && i + 1 < src.length() && src.charAt(i + 1) == '/') {
                while (i < src.length() && src.charAt(i) != '\n') i++;
                continue;
            }
            int col = i - lineStart + 1;
            // Numbers: integer or float
            if (Character.isDigit(c)) {
                int j = i;
                while (j < src.length() && Character.isDigit(src.charAt(j))) j++;
                boolean isFloat = false;
                if (j < src.length() && src.charAt(j) == '.') {
                    isFloat = true;
                    j++;
                    while (j < src.length() && Character.isDigit(src.charAt(j))) j++;
                }
                String text = src.substring(i, j);
                // NOTE: the branches return boxed Long / Double; merging them in a
                // ternary would auto-unbox and widen to double, silently losing the
                // integer type. We need to branch statementally to preserve it.
                Object num;
                if (isFloat) num = Double.parseDouble(text);
                else num = Long.parseLong(text);
                out.add(new Token(TokKind.NUMBER, text, num, line, col));
                i = j;
                continue;
            }
            // Strings: double-quoted with backslash escapes
            if (c == '"') {
                StringBuilder sb = new StringBuilder();
                int j = i + 1;
                while (j < src.length() && src.charAt(j) != '"') {
                    char ch = src.charAt(j);
                    if (ch == '\\' && j + 1 < src.length()) {
                        char esc = src.charAt(j + 1);
                        sb.append(switch (esc) {
                            case 'n' -> '\n';
                            case 't' -> '\t';
                            case 'r' -> '\r';
                            case '"' -> '"';
                            case '\\' -> '\\';
                            default -> esc;
                        });
                        j += 2;
                    } else {
                        sb.append(ch);
                        j++;
                    }
                }
                if (j >= src.length()) {
                    throw new IllegalStateException("unterminated string at line " + line);
                }
                String text = sb.toString();
                out.add(new Token(TokKind.STRING, text, text, line, col));
                i = j + 1;
                continue;
            }
            // Identifiers / keywords
            if (Character.isLetter(c) || c == '_') {
                int j = i;
                while (j < src.length() && (Character.isLetterOrDigit(src.charAt(j)) || src.charAt(j) == '_')) j++;
                String text = src.substring(i, j);
                if (KEYWORDS.contains(text)) {
                    out.add(new Token(TokKind.KEYWORD, text, text, line, col));
                } else {
                    out.add(new Token(TokKind.IDENT, text, text, line, col));
                }
                i = j;
                continue;
            }
            // Multi-char operators must be tried before single-char ones.
            String two = i + 1 < src.length() ? src.substring(i, i + 2) : "";
            if (Set.of("==", "!=", "<=", ">=", "&&", "||").contains(two)) {
                out.add(new Token(TokKind.OP, two, two, line, col));
                i += 2;
                continue;
            }
            if ("+-*/%=<>!(){},;".indexOf(c) >= 0) {
                String s = String.valueOf(c);
                out.add(new Token(TokKind.OP, s, s, line, col));
                i++;
                continue;
            }
            throw new IllegalStateException("unexpected character '" + c + "' at line " + line + ":" + col);
        }
        out.add(new Token(TokKind.EOF, "", null, line, 1));
        return out;
    }

    // ---------------------------------------------------------------------
    // Parser
    // ---------------------------------------------------------------------

    private final List<Token> toks;
    private int i;

    private Parser(List<Token> toks) { this.toks = toks; }

    /** Entry point: tokenize + parse a full program. */
    public static List<Ast.Rule> parse(String source) {
        return new Parser(tokenize(source)).parseProgram();
    }

    private Token peek() { return toks.get(i); }
    private Token peek(int k) { return toks.get(i + k); }

    private Token eat(TokKind kind) {
        Token t = toks.get(i);
        if (t.kind() != kind) {
            throw new IllegalStateException("expected " + kind + " at line " + t.line() + ":" + t.col() + ", got " + t.kind() + " '" + t.text() + "'");
        }
        i++;
        return t;
    }

    private Token eatKeyword(String kw) {
        Token t = toks.get(i);
        if (t.kind() != TokKind.KEYWORD || !kw.equals(t.text())) {
            throw new IllegalStateException("expected keyword '" + kw + "' at line " + t.line() + ":" + t.col() + ", got " + t.kind() + " '" + t.text() + "'");
        }
        i++;
        return t;
    }

    private boolean acceptKeyword(String kw) {
        Token t = toks.get(i);
        if (t.kind() == TokKind.KEYWORD && kw.equals(t.text())) { i++; return true; }
        return false;
    }

    private boolean acceptOp(String op) {
        Token t = toks.get(i);
        if (t.kind() == TokKind.OP && op.equals(t.text())) { i++; return true; }
        return false;
    }

    private void expectOp(String op) {
        if (!acceptOp(op)) {
            Token t = toks.get(i);
            throw new IllegalStateException("expected '" + op + "' at line " + t.line() + ":" + t.col() + ", got '" + t.text() + "'");
        }
    }

    // --- top level --------------------------------------------------------

    private List<Ast.Rule> parseProgram() {
        List<Ast.Rule> rules = new ArrayList<>();
        while (peek().kind() != TokKind.EOF) {
            rules.add(parseRule());
        }
        return rules;
    }

    private Ast.Rule parseRule() {
        eatKeyword("rule");
        String name = (String) eat(TokKind.STRING).value();
        String desc = "";
        long salience = 0;
        if (acceptKeyword("desc")) {
            desc = (String) eat(TokKind.STRING).value();
        }
        if (acceptKeyword("salience")) {
            long sign = acceptOp("-") ? -1 : 1;
            Object n = eat(TokKind.NUMBER).value();
            salience = sign * (n instanceof Long l ? l : ((Number) n).longValue());
        }
        eatKeyword("begin");
        List<Ast.Stmt> body = new ArrayList<>();
        while (!(peek().kind() == TokKind.KEYWORD && "end".equals(peek().text()))) {
            body.add(parseStatement());
        }
        eatKeyword("end");
        return new Ast.Rule(name, desc, salience, body);
    }

    // --- statements -------------------------------------------------------

    private Ast.Stmt parseStatement() {
        Token t = peek();
        if (t.kind() == TokKind.KEYWORD) {
            return switch (t.text()) {
                case "if" -> parseIf();
                case "return" -> { i++; yield new Ast.ReturnStmt(startsExpr() ? parseExpression() : null); }
                default -> throw new IllegalStateException("unexpected keyword '" + t.text() + "' at line " + t.line());
            };
        }
        if (t.kind() == TokKind.IDENT) {
            Token nxt = peek(1);
            if (nxt.kind() == TokKind.OP && "=".equals(nxt.text())) {
                String target = (String) eat(TokKind.IDENT).value();
                eat(TokKind.OP); // =
                return new Ast.AssignStmt(target, parseExpression());
            }
            if (nxt.kind() == TokKind.OP && "(".equals(nxt.text())) {
                return new Ast.CallStmt(parseCall());
            }
        }
        throw new IllegalStateException("unexpected token '" + t.text() + "' at line " + t.line() + ":" + t.col());
    }

    private Ast.IfStmt parseIf() {
        eatKeyword("if");
        Ast.Expr cond = parseExpression();
        expectOp("{");
        List<Ast.Stmt> thenBlock = parseBlockBody();
        expectOp("}");
        List<Ast.Stmt> elseBlock = List.of();
        if (acceptKeyword("else")) {
            expectOp("{");
            elseBlock = parseBlockBody();
            expectOp("}");
        }
        return new Ast.IfStmt(cond, thenBlock, elseBlock);
    }

    private List<Ast.Stmt> parseBlockBody() {
        List<Ast.Stmt> body = new ArrayList<>();
        while (!(peek().kind() == TokKind.OP && "}".equals(peek().text()))) {
            body.add(parseStatement());
        }
        return body;
    }

    private Ast.CallExpr parseCall() {
        String name = (String) eat(TokKind.IDENT).value();
        expectOp("(");
        List<Ast.Expr> args = new ArrayList<>();
        if (!(peek().kind() == TokKind.OP && ")".equals(peek().text()))) {
            args.add(parseExpression());
            while (acceptOp(",")) args.add(parseExpression());
        }
        expectOp(")");
        return new Ast.CallExpr(name, args);
    }

    // --- expressions (precedence climbing) --------------------------------

    private boolean startsExpr() {
        Token t = peek();
        if (t.kind() == TokKind.NUMBER || t.kind() == TokKind.STRING || t.kind() == TokKind.IDENT) return true;
        if (t.kind() == TokKind.KEYWORD) return "true".equals(t.text()) || "false".equals(t.text());
        if (t.kind() == TokKind.OP) return "-".equals(t.text()) || "!".equals(t.text()) || "(".equals(t.text());
        return false;
    }

    private Ast.Expr parseExpression() { return parseExpression(0); }

    private Ast.Expr parseExpression(int minPrec) {
        Ast.Expr lhs = parseUnary();
        while (true) {
            Token t = peek();
            if (t.kind() != TokKind.OP || !PRECEDENCE.containsKey(t.text())) break;
            int prec = PRECEDENCE.get(t.text());
            if (prec < minPrec) break;
            String op = t.text();
            i++;
            Ast.Expr rhs = parseExpression(prec + 1);
            lhs = new Ast.BinOpExpr(op, lhs, rhs);
        }
        return lhs;
    }

    private Ast.Expr parseUnary() {
        Token t = peek();
        if (t.kind() == TokKind.OP && ("-".equals(t.text()) || "!".equals(t.text()))) {
            i++;
            return new Ast.UnaryOpExpr(t.text(), parseUnary());
        }
        return parsePrimary();
    }

    private Ast.Expr parsePrimary() {
        Token t = peek();
        switch (t.kind()) {
            case NUMBER: i++; return new Ast.LiteralExpr(t.value());
            case STRING: i++; return new Ast.LiteralExpr(t.value());
            case KEYWORD:
                if ("true".equals(t.text())) { i++; return new Ast.LiteralExpr(Boolean.TRUE); }
                if ("false".equals(t.text())) { i++; return new Ast.LiteralExpr(Boolean.FALSE); }
                throw new IllegalStateException("unexpected keyword '" + t.text() + "' in expression");
            case IDENT:
                if (peek(1).kind() == TokKind.OP && "(".equals(peek(1).text())) return parseCall();
                i++;
                return new Ast.IdentExpr((String) t.value());
            case OP:
                if ("(".equals(t.text())) {
                    i++;
                    Ast.Expr e = parseExpression();
                    expectOp(")");
                    return e;
                }
                throw new IllegalStateException("unexpected operator '" + t.text() + "' at line " + t.line());
            default:
                throw new IllegalStateException("unexpected token kind " + t.kind());
        }
    }
}

File: ports/cpp/src/parser.cpp
// Hand-rolled lexer + recursive-descent parser for the C++ port.
// Grammar is identical to the Python and Java MVPs so rule files can be
// shared across ports when smoke-testing.

#include "gengine/parser.hpp"

#include <cctype>
#include <string>
#include <string_view>
#include <unordered_map>
#include <unordered_set>
#include <utility>

namespace gengine {
namespace {

// ---------------------------------------------------------------------------
// Tokens
// ---------------------------------------------------------------------------

enum class TokKind { Number, String, Ident, Op, Keyword, Eof };

struct Token {
    TokKind kind;
    std::string text;
    Value value;
    int line = 0;
    int col  = 0;
};

const std::unordered_set<std::string>& keywords() {
    static const std::unordered_set<std::string> kw{
        "rule", "desc", "salience", "begin", "end",
        "if", "else", "true", "false", "return"
    };
    return kw;
}

const std::unordered_map<std::string, int>& precedence() {
    static const std::unordered_map<std::string, int> p{
        {"||", 1}, {"&&", 2},
        {"==", 3}, {"!=", 3}, {"<", 3}, {"<=", 3}, {">", 3}, {">=", 3},
        {"+", 4}, {"-", 4},
        {"*", 5}, {"/", 5}, {"%", 5}
    };
    return p;
}

// ---------------------------------------------------------------------------
// Lexer
// ---------------------------------------------------------------------------

std::vector<Token> tokenize(const std::string& src) {
    std::vector<Token> out;
    std::size_t i = 0;
    int line = 1;
    std::size_t lineStart = 0;

    auto addOp = [&](std::string s, std::size_t len) {
        out.push_back({TokKind::Op, s, Value{s}, line, static_cast<int>(i - lineStart + 1)});
        i += len;
    };

    while (i < src.size()) {
        char c = src[i];
        if (c == '\n') { ++line; lineStart = ++i; continue; }
        if (std::isspace(static_cast<unsigned char>(c))) { ++i; continue; }
        // Line comment
        if (c == '/' && i + 1 < src.size() && src[i+1] == '/') {
            while (i < src.size() && src[i] != '\n') ++i;
            continue;
        }
        int col = static_cast<int>(i - lineStart + 1);

        // Numbers
        if (std::isdigit(static_cast<unsigned char>(c))) {
            std::size_t j = i;
            while (j < src.size() && std::isdigit(static_cast<unsigned char>(src[j]))) ++j;
            bool isFloat = false;
            if (j < src.size() && src[j] == '.') {
                isFloat = true;
                ++j;
                while (j < src.size() && std::isdigit(static_cast<unsigned char>(src[j]))) ++j;
            }
            std::string text = src.substr(i, j - i);
            Value v = isFloat ? Value{std::stod(text)}
                              : Value{static_cast<int64_t>(std::stoll(text))};
            out.push_back({TokKind::Number, text, v, line, col});
            i = j;
            continue;
        }

        // Strings
        if (c == '"') {
            std::string s;
            std::size_t j = i + 1;
            while (j < src.size() && src[j] != '"') {
                if (src[j] == '\\' && j + 1 < src.size()) {
                    char esc = src[j+1];
                    switch (esc) {
                        case 'n': s.push_back('\n'); break;
                        case 't': s.push_back('\t'); break;
                        case 'r': s.push_back('\r'); break;
                        case '"': s.push_back('"'); break;
                        case '\\': s.push_back('\\'); break;
                        default: s.push_back(esc); break;
                    }
                    j += 2;
                } else {
                    s.push_back(src[j++]);
                }
            }
            if (j >= src.size()) throw ParseError("unterminated string at line " + std::to_string(line));
            out.push_back({TokKind::String, s, Value{s}, line, col});
            i = j + 1;
            continue;
        }

        // Identifiers / keywords
        if (std::isalpha(static_cast<unsigned char>(c)) || c == '_') {
            std::size_t j = i;
            while (j < src.size() && (std::isalnum(static_cast<unsigned char>(src[j])) || src[j] == '_')) ++j;
            std::string text = src.substr(i, j - i);
            TokKind k = keywords().count(text) ? TokKind::Keyword : TokKind::Ident;
            out.push_back({k, text, Value{text}, line, col});
            i = j;
            continue;
        }

        // Two-char operators must be tried before one-char ones.
        if (i + 1 < src.size()) {
            std::string two = src.substr(i, 2);
            if (two == "==" || two == "!=" || two == "<=" || two == ">="
                || two == "&&" || two == "||") {
                addOp(two, 2);
                continue;
            }
        }
        if (std::string_view("+-*/%=<>!(){},;").find(c) != std::string_view::npos) {
            addOp(std::string(1, c), 1);
            continue;
        }

        throw ParseError("unexpected character '" + std::string(1, c) + "' at line "
                         + std::to_string(line) + ":" + std::to_string(col));
    }
    out.push_back({TokKind::Eof, "", {}, line, 1});
    return out;
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

class ParserImpl {
public:
    explicit ParserImpl(std::vector<Token> toks) : toks_(std::move(toks)) {}

    std::vector<Rule> parseProgram() {
        std::vector<Rule> rules;
        while (peek().kind != TokKind::Eof) rules.push_back(parseRule());
        return rules;
    }

private:
    std::vector<Token> toks_;
    std::size_t i_ = 0;

    const Token& peek(std::size_t k = 0) const { return toks_[i_ + k]; }

    const Token& eat(TokKind kind) {
        const auto& t = toks_[i_];
        if (t.kind != kind) fail("expected different kind", t);
        ++i_;
        return toks_[i_ - 1];
    }

    const Token& eatKeyword(const std::string& kw) {
        const auto& t = toks_[i_];
        if (t.kind != TokKind::Keyword || t.text != kw) fail("expected keyword '" + kw + "'", t);
        ++i_;
        return toks_[i_ - 1];
    }

    bool acceptKeyword(const std::string& kw) {
        if (peek().kind == TokKind::Keyword && peek().text == kw) { ++i_; return true; }
        return false;
    }

    bool acceptOp(const std::string& op) {
        if (peek().kind == TokKind::Op && peek().text == op) { ++i_; return true; }
        return false;
    }

    void expectOp(const std::string& op) {
        if (!acceptOp(op)) fail("expected '" + op + "'", peek());
    }

    [[noreturn]] void fail(const std::string& msg, const Token& t) {
        throw ParseError(msg + " at line " + std::to_string(t.line) + ":"
                         + std::to_string(t.col) + " (got '" + t.text + "')");
    }

    Rule parseRule() {
        eatKeyword("rule");
        std::string name = std::get<std::string>(eat(TokKind::String).value);
        Rule rule;
        rule.name = name;
        if (acceptKeyword("desc")) rule.desc = std::get<std::string>(eat(TokKind::String).value);
        if (acceptKeyword("salience")) {
            int64_t sign = acceptOp("-") ? -1 : 1;
            const Token& n = eat(TokKind::Number);
            int64_t val = std::visit([](auto&& v) -> int64_t {
                using T = std::decay_t<decltype(v)>;
                if constexpr (std::is_same_v<T, int64_t>) return v;
                else if constexpr (std::is_same_v<T, double>) return static_cast<int64_t>(v);
                else return 0;
            }, n.value);
            rule.salience = sign * val;
        }
        eatKeyword("begin");
        while (!(peek().kind == TokKind::Keyword && peek().text == "end")) {
            rule.body.push_back(parseStatement());
        }
        eatKeyword("end");
        return rule;
    }

    StmtPtr parseStatement() {
        const auto& t = peek();
        if (t.kind == TokKind::Keyword) {
            if (t.text == "if") return parseIf();
            if (t.text == "return") {
                ++i_;
                ExprPtr v;
                if (startsExpr()) v = parseExpression();
                return std::make_shared<Stmt>(Stmt{ReturnStmt{v}});
            }
            fail("unexpected keyword in statement", t);
        }
        if (t.kind == TokKind::Ident) {
            const auto& nxt = peek(1);
            if (nxt.kind == TokKind::Op && nxt.text == "=") {
                std::string target = t.text;
                ++i_; ++i_; // IDENT =
                return std::make_shared<Stmt>(Stmt{AssignStmt{target, parseExpression()}});
            }
            if (nxt.kind == TokKind::Op && nxt.text == "(") {
                return std::make_shared<Stmt>(Stmt{CallStmt{parseCall()}});
            }
        }
        fail("unexpected token in statement", t);
    }

    StmtPtr parseIf() {
        eatKeyword("if");
        ExprPtr cond = parseExpression();
        expectOp("{");
        std::vector<StmtPtr> thenBlock;
        while (!(peek().kind == TokKind::Op && peek().text == "}"))
            thenBlock.push_back(parseStatement());
        expectOp("}");
        std::vector<StmtPtr> elseBlock;
        if (acceptKeyword("else")) {
            expectOp("{");
            while (!(peek().kind == TokKind::Op && peek().text == "}"))
                elseBlock.push_back(parseStatement());
            expectOp("}");
        }
        return std::make_shared<Stmt>(Stmt{IfStmt{cond, std::move(thenBlock), std::move(elseBlock)}});
    }

    CallExpr parseCall() {
        std::string name = eat(TokKind::Ident).text;
        expectOp("(");
        CallExpr c; c.name = name;
        if (!(peek().kind == TokKind::Op && peek().text == ")")) {
            c.args.push_back(parseExpression());
            while (acceptOp(",")) c.args.push_back(parseExpression());
        }
        expectOp(")");
        return c;
    }

    bool startsExpr() const {
        const auto& t = peek();
        if (t.kind == TokKind::Number || t.kind == TokKind::String || t.kind == TokKind::Ident) return true;
        if (t.kind == TokKind::Keyword) return t.text == "true" || t.text == "false";
        if (t.kind == TokKind::Op) return t.text == "-" || t.text == "!" || t.text == "(";
        return false;
    }

    ExprPtr parseExpression(int minPrec = 0) {
        ExprPtr lhs = parseUnary();
        while (true) {
            const auto& t = peek();
            if (t.kind != TokKind::Op || !precedence().count(t.text)) break;
            int prec = precedence().at(t.text);
            if (prec < minPrec) break;
            std::string op = t.text;
            ++i_;
            ExprPtr rhs = parseExpression(prec + 1);
            lhs = std::make_shared<Expr>(Expr{BinOpExpr{op, lhs, rhs}});
        }
        return lhs;
    }

    ExprPtr parseUnary() {
        const auto& t = peek();
        if (t.kind == TokKind::Op && (t.text == "-" || t.text == "!")) {
            std::string op = t.text;
            ++i_;
            return std::make_shared<Expr>(Expr{UnaryOpExpr{op, parseUnary()}});
        }
        return parsePrimary();
    }

    ExprPtr parsePrimary() {
        const auto& t = peek();
        switch (t.kind) {
            case TokKind::Number:
                ++i_;
                return std::make_shared<Expr>(Expr{LiteralExpr{t.value}});
            case TokKind::String:
                ++i_;
                return std::make_shared<Expr>(Expr{LiteralExpr{t.value}});
            case TokKind::Keyword:
                if (t.text == "true")  { ++i_; return std::make_shared<Expr>(Expr{LiteralExpr{Value{true}}}); }
                if (t.text == "false") { ++i_; return std::make_shared<Expr>(Expr{LiteralExpr{Value{false}}}); }
                fail("unexpected keyword in expression", t);
            case TokKind::Ident:
                if (peek(1).kind == TokKind::Op && peek(1).text == "(")
                    return std::make_shared<Expr>(Expr{parseCall()});
                ++i_;
                return std::make_shared<Expr>(Expr{IdentExpr{t.text}});
            case TokKind::Op:
                if (t.text == "(") {
                    ++i_;
                    ExprPtr e = parseExpression();
                    expectOp(")");
                    return e;
                }
                fail("unexpected operator in expression", t);
            case TokKind::Eof:
                fail("unexpected end of input", t);
        }
        fail("unreachable", t);
    }
};

} // namespace

std::vector<Rule> parse(const std::string& source) {
    return ParserImpl(tokenize(source)).parseProgram();
}

} // namespace gengine
